### ISSUE TITLE
Extend `@agent` macro to allow inheritance

### ISSUE THREAD

**Issue by fbanning** [JuliaDynamics/Agents.jl#651]
It would be great to extend the `@agent` macro to allow for field inheritance.

```julia
@agent Animal{I} GridAgent{2} begin
    age::I
end

@agent Bird{F} Animal{I} begin
    wingspread::F
end
```

The current implementation errors out when trying to define the agent subtype:

```julia
julia> @agent Bird{F} Animal{I} begin
           wingspread::F
       end
ERROR: LoadError: UndefVarError: Animal not defined
Stacktrace:
 [1] top-level scope
   @ none:1
 [2] eval(m::Module, e::Any)
   @ Core ./boot.jl:373
 [3] var"@agent"(__source__::LineNumberNode, __module__::Module, name::Any, base::Any, fields::Any)
   @ Agents ~/.julia/packages/Agents/sza9W/src/core/agents.jl:62
in expression starting at REPL[8]:1
```

**Comment by Datseris** [comment]
I love this suggestion and it further helps with my plans of making `@agents` the only recommended way to make agent types in the documentation. Too bad my metaprogramming skills are nonexistent to solve it :D

**Comment by Datseris** [comment]
Hm, isn't it as simply as changing this line: 

https://github.com/JuliaDynamics/Agents.jl/blob/main/src/core/agents.jl#L62

It tries to find the second type in the `Agents` module, but it should try the module the code is executed in as well, right?

**Comment by fbanning** [comment]
Another thing to look at would be the fact that types created with the `@agent` macro do not seem to be a subtype of `GridAgent`

```julia
julia> Animal <: GridAgent
false
```

but are indeed a subtype of `AbstractAgent`

```julia
julia> Animal <: AbstractAgent
true
```

According to how the macro is structured, I would expect the type tree to look something like

```julia
Animal <: GridAgent <: AbstractAgent
```

**Comment by fbanning** [comment]
This might be as simple as changing [this line](https://github.com/JuliaDynamics/Agents.jl/blob/main/src/core/agents.jl#L66) to something like

```julia
res = :(mutable struct $(esc(name)) <: base end
```

**Comment by Datseris** [comment]
that's exactly what I was about to write, although I think you'll need `$(base)`.

**Comment by Datseris** [comment]
Ah no, that is not possible, now I remember. `GridAgent{2}` is a concrete type, and therefore it is not possible for it to have subtypes.

**Comment by Datseris** [comment]
You'll have to leave with `YourType <: AbstractAgent` only unfortunately. The `<:` operator can't be hacked in Julia. I guess that's the reason TyOOP also doesn't have intuitive behavior with `<:`.

**Comment by Datseris** [comment]
In your example above `Animal` would also be an instantiable concrete type, and therefore cannot have subtypes.

**Comment by Datseris** [comment]
> That's really shitty

But also unavoidable with standard Julia code. Only concrete types can have fields, and concrete types cannot have subtypes. That's what all these OOP packages do: they make some weird abstract types wrapping concrete types to make it possible.

Then again, why would you need different agent types to subtype each other? What is a potential application of that?

If you care about dispatch, you can do 
```
Animal = Union{Bird, Flock}
```
but, perhaps here duck typing is the best solution (i.e., never specify types unless in the lowest level methods)

**Comment by Datseris** [comment]
My suggestion would be to write `get_older(x)` instead of `get_older(x::Mammal)`. `Mammal` doesn't have reason to exist yet, you can use ducktyping here. The only scenario I can imagine with intermediate abstract types suich as `Mammal` is if you have several of them, such as `Mammal` but also `Insect`. But in such a case, I imagine the source code for the ABM stepping etc would be so massive, that the extra two lines that define the two `Union` types would really not be a big deal.

**Comment by fbanning** [comment]
That's true but doing it with the Union and without a proper type relation between them, it doesn't really provide a promise that both subtypes have the fields of the supertype `Mammal` (which is specifically why I asked about the possibility of inheritance by extending the `@agent` macro). Anyways, this is something to keep in mind because maybe at some point it will be possible (or maybe with really advanced metaprogramming skills that's already the case now and we simply don't know).

**Comment by Datseris** [comment]
> That's true but doing it with the Union and without a proper type relation between them, it doesn't really provide a promise that both subtypes have the fields of the supertype Mammal (which is specifically why I asked about the possibility of inheritance by extending the @agent macro). 

Subtyping relations anyways provide no promise whatsoever on the fields of types within the Julia language. If types `A, B` satisfy `A <: B` you still can say nothing about the fields of either, at least within the formal framework of julia.

So the question is not a matter of subtyping or not, it's about whether the _macro_ `@agent` (or whatever macro from some OOP package) can promise you that the way you create your new type it has the fields of some other type. You will never get this information by subtyping, at least in current versions of Julia. :(

**Comment by fbanning** [comment]
Hm, you seem to misunderstand me. I'm fully aware of these facts. Hence why I proposed to (i) extend the `@agent` macro with inheritance capabilities while (ii) trying to ensure proper subtyping inside Julia's type system. If that's not possible with the macro then that's a shame but likely nothing we can do anything about.

**Comment by fbanning** [comment]
Still, I think, the whole idea of this issue might be worth it, even without proper Julia subtyping. It would still enable users to easily build structs on top of existing agent types and extend them to their liking with some extra fields.

**Comment by Datseris** [comment]
Maybe I did misunderstand, but what I was trying to make clear is that to have both (i) and (ii) at the same time is impossible in the current version of julia, irrespectively of macros, because only concrete types can have fields (which is what you want to inherit) but only abstract types can be subtyped. Same reason why in TyOOP.jl you don't have `Snake <: Animal ` as you pointed out in the discourse post. 

So, TLDR, don't expect (i) and (ii) happening at the same time in current Julia... :(
